lib/Sema/CSDiagnostics.cpp

```diff
@@ -1000,23 +1029,41 @@ bool MemberAccessOnOptionalBaseFailure::diagnoseAsError() {
   auto unwrappedBaseType = baseType->getOptionalObjectType();
   if (!unwrappedBaseType)
     return false;
+  
+  auto baseSourceRange = getMemberBaseSourceRange();
+  
+  emitDiagnosticAt(baseSourceRange.End, diag::optional_base_not_unwrapped,
+                   baseType, Member, unwrappedBaseType);
+
+  auto componentPathElt =
+      locator->getLastElementAs<LocatorPathElt::KeyPathComponent>();
+  if (componentPathElt && componentPathElt->getIndex() == 0) {
+    // For members where the base type is an optional key path root
+    // let's emit a tailored note suggesting to remove the '?' and a
+    // fix to replace the optional type with its unwrapped type.
+    if (auto *nominalDecl = unwrappedBaseType->getAnyNominal()) {
+      auto *keyPathExpr = castToExpr<KeyPathExpr>(getAnchor());
+      auto unwrappedName = nominalDecl->getBaseName();
+      emitDiagnostic(diag::optional_base_remove_optional_for_keypath_root,
+                     baseType)
+          .fixItReplace(keyPathExpr->getRootType()->getSourceRange(),
+                        unwrappedName.getIdentifier().str());
+    }
+    
+  } else {
+    // FIXME: It would be nice to immediately offer "base?.member ?? defaultValue"
+    // for non-optional results where that would be appropriate. For the moment
+    // always offering "?" means that if the user chooses chaining, we'll end up
+    // in MissingOptionalUnwrapFailure:diagnose() to offer a default value during
+    // the next compile.
+    emitDiagnostic(diag::optional_base_chain, Member)
```

Comment:
Note that these diagnostics will use the source location provided by `FailureDiagnostic::getLoc`, which will use the entire key path's location. I think it would make sense to override `getLoc` for `MemberAccessOnOptionalBaseFailure` and have it return the end loc of the source range we compute.